### PR TITLE
Detect stream operators for pretty compile-time errors

### LINKED ISSUE #3: Add static_assert for better error messages
Obviously, we can not create a `painless::Parameter<T>` for any `T`. We should think about the proper constraint and add a `static_assert` such that users will see a better error message.

For now, we probably require existence of `ostream& operator<<` and `istream& operator>>` for "serialization" and "deserialization".

### PR DESCRIPTION
**Pull request by hmenke**
Fixes #3 

Basically copy-pasted from https://github.com/sharkdp/dbg-macro

### CODE REVIEW

File: include/painless/parameter.h [hunk 1]
```diff
@@ -31,38 +88,66 @@ inline std::string get_base_path() {
...
+inline std::string to_string(const bool& value, std::true_type) {
   std::stringstream ss;
   ss << std::boolalpha << value;
   return ss.str();
 }
 
+template <typename T>
+inline std::string to_string(const T& value) {
+  return to_string(value,
+                   typename detail::has_ostream_operator<const T&>::type{});
```
Comment (sharkdp) [inline include/painless/parameter.h]: Can you remind me why tag dispatch is required/preferred here? Couldn't we simply add a
```c++
static_assert(detail::has_istream_operator<const T&>::value,
              "Type does not support the << ostream operator"); 
```
in the class scope of `Parameter<T>`?

Comment (hmenke) [reply include/painless/parameter.h]: That would also be possible.  I thought that having check at the point of use is maybe more reusable.

File: include/painless/parameter.h [hunk 2]
```diff
@@ -31,38 +88,66 @@ inline std::string get_base_path() {
...
+}
+
 }  // namespace parser
 
 namespace printer {
 
 template <typename T>
-inline std::string to_string(const T& value) {
+inline std::string to_string(const T&, std::false_type) {
+  static_assert(detail::has_istream_operator<const T&>::value,
```
Comment (sharkdp) [inline include/painless/parameter.h]: This should be `has_ostream_operator<…>`.

The `to_string`/`from_string` methods are internal implementation details, while the `Parameter<T>` class is part of the public API.

I think I would therefore prefer the simpler solution where we add two separate `static_assert`s for both `has_ostream_operator` and `has_istream_operator` to the `Parameter` class template. Having an instance of `Parameter<T>` for `T` without any of these operators is not sensible/possible/useful, so we are not artificially restricting the usage.

**Comment by sharkdp** [comment]
Thank you!

**Pull request merged**